### DISCUSSION TITLE
Error: Unknown command: cask

### DISCUSSION THREAD

**Discussion by mavabene** [Homebrew/discussions#902] [Casks]
I'm new to programming.  I can code some in python and Matlab but I don't understand command line interfaces or what is going on behind the scenes.

I was doing a tutorial to learn about Mac Terminal and they had me install homebrew.  Then Cask.  I tried to run "brew cask help" and I get "Error: Unknown command : cask".  I have googled and tried things and added paths and brew doctored and upgraded and reinstalled and cleared directories and cleaned up and I consistently just get this error.  What is going on??

Any help would be greatly appreciated!  And help with basic explanations even more so!  Thank you.

brew doctor: "Your system is ready to brew." 
macOS Big Sur Version 11.2.1

**Comment by carlocab** [comment]
Your tutorial is outdated. `cask` is no longer a `brew` command. When you want to install a Cask, you just do `brew install` or `brew install --cask` instead of `brew cask install`.

**Reply by mavabene** [reply]
Lol, omg.  I spent hours trying to solve something that wasn't a problem?  Thank you!

**Reply by carlocab** [reply]
Possibly. If you need to know how to do something with `brew`, you can always use `brew help`, or read the manual with `man brew`. `brew help <command>` will also show you a help message for a command (e.g. `brew help upgrade`).

**Reply by mavabene** [reply]
Thanks!

**Reply by jacob-greenfield** [reply]
I just had the same exact experience, spent a few frustrating hours digging only to realize that the command has been apparently removed as part of Homebrew 3.0.0 (which I didn't even know was a thing; I just ran `brew upgrade` like I normally do every so often). But either way, there wasn't any mention of this in the [release notes](https://brew.sh/2021/02/05/homebrew-3.0.0/), or really anywhere that I can find. Even the PR removing the `cask` command (Homebrew/brew#10397) didn't mention it. This was a breaking change, but I don't think I ever saw a deprecation notice, it just suddenly stopped working after I ran `brew upgrade`. Moving forward, is there any way to get a heads up for breaking changes like this, before they land?

Back to the `cask` issue, at the very least, might it be a good idea to change the error message from `Error: Unknown command: cask` to something like this: ``Error: The `cask` command has been removed. Use `brew install --cask` instead``?

**Reply by SMillerDev** [reply]
It was part of [brew 2.6](https://brew.sh/2020/12/01/homebrew-2.6.0/) and had been issuing the depreciation notice you suggested for roughly 3 months when it was removed in 3.0

**Reply by TKaushik15** [reply]
> brew help upgrade

I am trying to install pyspark on my mac. i have installed homebrew using the command 
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"

I tried installing install Java as Spark is written in Scala using the following command
brew cask install java

I am getting Error: Unknown command: cask in the output. 

What should I do now. Kindly help....

**Reply by SMillerDev** [reply]
`brew install --cask java` or `brew install openjdk`

**Reply by tlk** [reply]
> This was a breaking change, but I don't think I ever saw a deprecation notice, it just suddenly stopped working

The outcome of this is very predictable and it does not put Homebrew in a good light.

Not only do the breaking changes affect existing users negatively, it also prevents new users from adopting Homebrew considering the vast amount of existing tutorials, stackoverflow answers, blog posts, etc. that are now broken.

**Reply by artforlife** [reply]
Whats the substitute for `brew cask cleanup`?

**Reply by SMillerDev** [reply]
https://docs.brew.sh/Manpage#cleanup-options-formulacask-

**Reply by malikonakci** [reply]
thanks i think you solved my problem too.

**Reply by phoenixeliot** [reply]
I think it would be best to keep the deprecation notice up for much longer, maybe forever — there will still be old tutorials on the internet for a very long time, and it will save people a google search.

**Reply by SMillerDev** [reply]
We didn't keep Constantinople the capital of the Ottoman empire because someone already wrote it down. If people don't update their blogs that's not on homebrew to fix.

**Reply by phoenixeliot** [reply]
Hrm. Makes for bad UX though for a lot of people's use cases. Is it really not worth the extra couple lines of code?

The problem makes me think of Link Rot, where websites become much harder to use and less useful over time if the URL scheme changes every now and then, and a lot of user pain can be prevented by at least making the old URLs do something to help the user find what they were looking for (if the URL scheme really does have to change).

**Reply by phoenixeliot** [reply]
Seems like something like this is all it would take to improve that UX noticeably at nearly zero code cost:
```
  def command
    args = command_args.parse

    args.named.each do |cmd|
      path = Commands.path(cmd)
      odie "cask command is deprecated. Use brew install --cask instead." if cmd == "cask"
      odie "Unknown command: #{cmd}" unless path
      puts path
    end
  end
```
at https://github.com/Homebrew/brew/blob/3476ca7b8ff889ef61e65783d35d1b2073bd4012/Library/Homebrew/dev-cmd/command.rb#L23-L31

**Reply by Matt-Hill-83** [reply]
maybe brew could provide a more useful error message when people inevitably use the deprecated syntax

**Reply by D1-4N4** [reply]
Thanks so much! Saved myself from. another headache

**Reply by rhzs** [reply]
@carlocab we should have proper error message. Not everyone has the same knowledge like yours or the homebrew team when deprecating something.  Not just `Error: Unknown command: cask` . That does not help anything. A better version will be `Do you mean brew install --cask XXX?`

**Reply by SMillerDev** [reply]
> A better version will be Do you mean brew install --cask XXX?

There was a similar message during the depreciation around a year ago. Be won't keep old code around forever.

**Reply by phoenixeliot** [reply]
I really don't understand how it's reasonable to just pretend like all the tutorials/guide/documentation that have the old command should just be treated like they don't exist or don't matter. Old documentation that is otherwise valid except for this command is going to keep existing forever, that's the nature of these things. Many people will keep running into this problem and be at a loss because they "should" have somehow known about this deprecation, but will only find out about it after googling for this error and having to find this discussion thread.

**Reply by SMillerDev** [reply]
But what are the alternatives in your eyes? Forcing volunteers to maintain code they want to remove indefinitely because someone wrote a tutorial about it isn't a good solution either.

Homebrew did it's part by deprecating the command and changing all our documents. It's now up to other authors to maintain their part.

**Reply by phoenixeliot** [reply]
I really don't think the single-line addition I suggested above is much of a maintenance headache. Is it?
I suspect the time spent reading this thread so far would be comparable if not more than the maintenance load of adding that one line and leaving it there.

**Reply by carlocab** [reply]
If it really is that simple, then please make a pull request. It may not be accepted, but if it takes as little effort as you say it does, then  opening a PR shouldn't hurt.

That said, I also don't understand how it's reasonable for the authors of software to be made responsible for documentation they never wrote or endorsed.

**Reply by phoenixeliot** [reply]
It feels like we're coming at this from two different motivations — My motivation here is just pragmatism. I think it will make a lot of people's lives easier on the order of saving them an hour or two of fighting with this command, and that's a lot of time wasted compared to the effort of having a good error message, and that makes having a good error message a good idea that people will get value out of.

Basically: Sure, maybe you shouldn't *have* to, but it's a nice thing to do for all those people.

I'll submit a PR.

**Reply by phoenixeliot** [reply]
I'm having trouble finding the instructions for building brew from source. Is there a guide to that that I'm failing to find? I found many other related guides in the repo but not those steps, and this is not very googleable (all results are about using homebrew to build other things from source).

Same question for running any required tests once I have it built with my modifications. I found some docs on adding tests.

**Reply by SMillerDev** [reply]
Brew is written in ruby so it isn't build. Any changes are immediately usable.

You can use `brew tests` to run internal tests.

**Reply by phoenixeliot** [reply]
Submitted https://github.com/Homebrew/brew/pull/12708 :)

**Reply by phoenixeliot** [reply]
Aside: It took me a bit to realize that anywhere the docs said to run `brew <something>` I actually needed to run `./bin/brew <something>`

**Reply by carlocab** [reply]
> Aside: It took me a bit to realize that anywhere the docs said to run `brew <something>` I actually needed to run `./bin/brew <something>`

No, the docs definitely intend for you to run `brew <something>`. There's something off about your config if you need to run `./bin/brew <something>` instead.

**Reply by phoenixeliot** [reply]
Dunno what to say. `brew` points to the one in my normal PATH, even when in the repo folder.

**Reply by nandahkrishna** [reply]
This is intentional – modifications to Homebrew code (for a potential PR) are made in your existing Homebrew installation/source tree, in a new branch. So you needn't separately clone the Homebrew repository at all. See https://docs.brew.sh/How-To-Open-a-Homebrew-Pull-Request#core-brew-code-related-pull-request (specifically, `cd "$(brew --repository)"`). Once you've made the changes you want to, running `brew <something>` should work just fine.

**Reply by phoenixeliot** [reply]
Aah, guess I just missed that. Thanks!

**Comment by gustawdaniel-acaisoft** [comment]
So this is outdated too?

https://www.cprime.com/resources/blog/docker-on-mac-with-homebrew-a-step-by-step-tutorial/

It is my first day on mac

**Reply by SMillerDev** [reply]
Yeah, anything that mentions `brew cask` is outdated.

**Reply by carlocab** [reply]
The `brew cask install` part is outdated. Generally, any command that is written as
```
brew cask <command>
```
can/should be run as
```
brew <command> --cask
```
